python/phase-1.md

```diff
@@ -0,0 +1,441 @@
+
+    @property
+    def cancellation(self) -> asyncio.Task:
+        pass
+
+
+# Adapter that extracts info() only when available in context
+class LoggerAdapter(logging.LoggerAdapter):
+    pass
+
+
+_logger = LoggerAdapter(logging.getLogger(__name__))
+
+# Logger that activities can use w/ activity context info instead of making
+# their own (though making your own is recommended)
+def logger() -> logging.Logger:
+    return _logger
+
+
+def heartbeat(*details: Any) -> None:
+    pass
+```
+
+Notes:
+
+* These all uses `contextvars` to get activity info or throw otherwise
+  * We will make the contextual state open to caller injection to support test mocking. The details are undecided atm.
+
+### Activity Definition/Registration
+
+```python
+from temporalio import activity
+
+# This adds activity details if they are present. Maybe we should have a
+# temporalio.LoggerAdapter that works for activities _and_ workflows?
+my_activity_logger = activity.LoggerAdapter(logging.getLogger(__name__))
+
+
+def simple_activity(arg1, arg2):
+    # Uses contextvars internally, which means it'll fail in another context but
+    # we will provide the ability to set this for testing situations
+    attempts = activity.info().attempts
+    # This logger has an adapter to include custom details but still named after
+    # this module which is a common pattern
+    my_activity_logger.debug("Attempt %s", attempts)
+    # This logger, on the activity module, uses the name of the activity module.
+    # So that's why we offer either the adapter or the logger.
+    activity.logger().debug("Attempt %s", attempts)
+
+
+async def another_activity(arg1):
+    # Showing heartbeat and cancellation. In Python 3.7 we cannot give a
+    # specific error to Task/Future.cancel (only in 3.9). So activity
+    # cancellations appear like any other asyncio cancellations (which may be a
+    # good thing anyways).
+    try:
+        while True:
+            await asyncio.sleep(1)
+            activity.heartbeat("some details")
+    except asyncio.CancelledError as exc:
+        raise RuntimeError("someone cancelled me!") from exc
+
+
+class MyCallableClass:
+    def __call__(self):
+        pass
+
+
+class MyActivities:
+    def activity1():
+        pass
+
+
+def create_my_worker() -> temporalio.Worker:
+    return temporalio.Worker(
+        temporalio.Core.default_local(),
+        taskqueue="MyTaskQueue",
+        activities={
+            # The names can be any value
+            "simple_activity": simple_activity,
+            # Can use async function
+            "another_activity": another_activity,
+            # Can use callable class
+            "callable_class": MyCallableClass(),
+            # Can use method
+            "activity1_method": MyActivities().activity1,
+            # Can use lambda
+            "lambda": lambda arg1: arg1,
+            # Fails because not callable
+            "not_callable": MyActivities(),
+        },
+    )
+```
+
+Notes:
```

Comment:
Need to document our discussions around why we removed activity decorator and why we require a mapping w/ explicit activity name

Reply:
Documented